File: ts/src/discovery_cache.ts
/**
 * DiscoveryCache — TTL + LRU cache for DiscoveryEntry objects.
 *
 * Mirrors Python roar_sdk/discovery_cache.py exactly.
 * Uses a Map (insertion-order preserved in JS) to implement LRU:
 *   - On hit:  delete + re-insert (moves to end / most-recent)
 *   - On evict: remove via Map.keys().next() (removes front / least-recent)
 */

import type { DiscoveryEntry } from "./types.js";

// ---------------------------------------------------------------------------
// Internal cache entry (not exported)
// ---------------------------------------------------------------------------

interface _CacheEntry {
  entry: DiscoveryEntry;
  cached_at: number; // unix timestamp (seconds, float)
  ttl: number;       // seconds
}

function _isExpired(e: _CacheEntry): boolean {
  return Date.now() / 1000 - e.cached_at > e.ttl;
}

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface DiscoveryCacheStats {
  size: number;
  max_entries: number;
  hits: number;
  misses: number;
  hit_rate: number;
}

// ---------------------------------------------------------------------------
// DiscoveryCache
// ---------------------------------------------------------------------------

export class DiscoveryCache {
  private _cache: Map<string, _CacheEntry>;
  private _max_entries: number;
  private _default_ttl: number;
  private _hits: number;
  private _misses: number;

  constructor(max_entries = 1000, default_ttl = 300.0) {
    this._cache = new Map();
    this._max_entries = max_entries;
    this._default_ttl = default_ttl;
    this._hits = 0;
    this._misses = 0;
  }

  /** Look up a DiscoveryEntry by DID. Returns undefined on miss or expiry. */
  get(did: string): DiscoveryEntry | undefined {
    const e = this._cache.get(did);
    if (e === undefined || _isExpired(e)) {
      this._misses++;
      if (e !== undefined) {
        // expired — remove it
        this._cache.delete(did);
      }
      return undefined;
    }
    // LRU: move to end (most-recently used)
    this._cache.delete(did);
    this._cache.set(did, e);
    this._hits++;
    return e.entry;
  }

  /** Insert or update a DiscoveryEntry in the cache. */
  put(entry: DiscoveryEntry, ttl?: number): void {
    const did = entry.agent_card.identity.did;
    const resolvedTtl = ttl !== undefined ? ttl : this._default_ttl;

    // If key already present, remove first so re-insert lands at the end
    this._cache.delete(did);
    this._cache.set(did, {
      entry,
      cached_at: Date.now() / 1000,
      ttl: resolvedTtl,
    });

    // Evict LRU entries (front of Map) while over capacity
    while (this._cache.size > this._max_entries) {
      const oldest = this._cache.keys().next().value;
      if (oldest !== undefined) {
        this._cache.delete(oldest);
      }
    }
  }

  /** Remove a single entry. Returns true if it existed. */
  invalidate(did: string): boolean {
    return this._cache.delete(did);
  }

  /** Remove all cached entries. Stats counters are preserved (matches Python). */
  clear(): void {
    this._cache.clear();
  }

  /**
   * Search for entries that advertise a given capability.
   * Checks both agent_card.identity.capabilities and agent_card.skills.
   * Expired entries are evicted before searching.
   */
  search(capability: string): DiscoveryEntry[] {
    this._evict_expired();
    const results: DiscoveryEntry[] = [];
    for (const e of this._cache.values()) {
      const card = e.entry.agent_card;
      if (
        card.identity.capabilities.includes(capability) ||
        card.skills.includes(capability)
      ) {
        results.push(e.entry);
      }
    }
    return results;
  }

  /** Number of entries currently in the cache. */
  get size(): number {
    return this._cache.size;
  }

  /** Cache statistics snapshot. */
  get stats(): DiscoveryCacheStats {
    const total = this._hits + this._misses;
    return {
      size: this._cache.size,
      max_entries: this._max_entries,
      hits: this._hits,
      misses: this._misses,
      hit_rate: total === 0 ? 0.0 : this._hits / total,
    };
  }

  /** Remove all expired entries from the cache. */
  private _evict_expired(): void {
    for (const [did, e] of this._cache.entries()) {
      if (_isExpired(e)) {
        this._cache.delete(did);
      }
    }
  }
}

File: python/src/roar_sdk/discovery_cache.py
# -*- coding: utf-8 -*-
"""TTL + LRU discovery cache for agent lookup results.

Wraps any agent directory with a fast in-memory cache layer.
Entries are evicted when they expire (TTL) or when the cache is
at capacity (LRU eviction of the oldest entry).

Usage::

    cache = DiscoveryCache(max_entries=500, default_ttl=300.0)

    # On lookup
    entry = cache.get(did)
    if entry is None:
        entry = directory.lookup(did)
        if entry:
            cache.put(entry)
"""

from __future__ import annotations

import time
from collections import OrderedDict
from dataclasses import dataclass
from typing import Dict, List, Optional

from .types import DiscoveryEntry


@dataclass
class _CacheEntry:
    entry: DiscoveryEntry
    cached_at: float
    ttl: float

    @property
    def expired(self) -> bool:
        return time.time() - self.cached_at > self.ttl


class DiscoveryCache:
    """TTL + LRU discovery cache.

    Args:
        max_entries: Maximum number of entries before LRU eviction.
        default_ttl: Default time-to-live in seconds.
    """

    def __init__(self, max_entries: int = 1000, default_ttl: float = 300.0) -> None:
        self._max_entries = max_entries
        self._default_ttl = default_ttl
        self._cache: OrderedDict[str, _CacheEntry] = OrderedDict()
        self._hits = 0
        self._misses = 0

    def get(self, did: str) -> Optional[DiscoveryEntry]:
        """Look up a cached entry by DID. Returns None on miss or expiry."""
        ce = self._cache.get(did)
        if ce is None:
            self._misses += 1
            return None
        if ce.expired:
            del self._cache[did]
            self._misses += 1
            return None
        self._cache.move_to_end(did)
        self._hits += 1
        return ce.entry

    def put(self, entry: DiscoveryEntry, ttl: Optional[float] = None) -> None:
        """Cache a discovery entry.

        Args:
            entry: The discovery entry to cache.
            ttl: Override TTL in seconds (uses default_ttl if omitted).
        """
        did = entry.agent_card.identity.did
        self._cache[did] = _CacheEntry(
            entry=entry,
            cached_at=time.time(),
            ttl=ttl if ttl is not None else self._default_ttl,
        )
        self._cache.move_to_end(did)
        while len(self._cache) > self._max_entries:
            self._cache.popitem(last=False)

    def invalidate(self, did: str) -> bool:
        """Remove a specific entry from the cache."""
        if did in self._cache:
            del self._cache[did]
            return True
        return False

    def clear(self) -> None:
        """Clear all cached entries."""
        self._cache.clear()

    def search(self, capability: str) -> List[DiscoveryEntry]:
        """Search cached entries by capability string."""
        self._evict_expired()
        return [
            ce.entry
            for ce in self._cache.values()
            if capability in ce.entry.agent_card.identity.capabilities
               or capability in ce.entry.agent_card.skills
        ]

    @property
    def size(self) -> int:
        return len(self._cache)

    @property
    def stats(self) -> Dict:
        total = self._hits + self._misses
        return {
            "size": self.size,
            "max_entries": self._max_entries,
            "hits": self._hits,
            "misses": self._misses,
            "hit_rate": self._hits / total if total > 0 else 0.0,
        }

    def _evict_expired(self) -> None:
        expired = [did for did, ce in self._cache.items() if ce.expired]
        for did in expired:
            del self._cache[did]
